presto-main/src/main/java/io/prestosql/sql/planner/QueryPlanner.java

```diff
@@ -857,15 +857,15 @@ private PlanBuilder distinct(PlanBuilder subPlan, QuerySpecification node)
 
     private PlanBuilder sort(PlanBuilder subPlan, Query node)
     {
-        return sort(subPlan, node.getOrderBy(), node.getLimit(), analysis.getOrderByExpressions(node));
+        return sort(subPlan, node.getOrderBy(), analysis.getLimit(node), analysis.getOrderByExpressions(node));
     }
 
     private PlanBuilder sort(PlanBuilder subPlan, QuerySpecification node)
     {
-        return sort(subPlan, node.getOrderBy(), node.getLimit(), analysis.getOrderByExpressions(node));
+        return sort(subPlan, node.getOrderBy(), analysis.getLimit(node), analysis.getOrderByExpressions(node));
     }
 
-    private PlanBuilder sort(PlanBuilder subPlan, Optional<OrderBy> orderBy, Optional<String> limit, List<Expression> orderByExpressions)
+    private PlanBuilder sort(PlanBuilder subPlan, Optional<OrderBy> orderBy, Long limit, List<Expression> orderByExpressions)
```

Comment:
If you associate the limit value in the analysis with the node for the clause (FetchFirst or Limit) instead of the Query/QuerySpecification, this method can take an `Optional<Node> limit` and the code below can just say:

```java
if (limit.isPresent()) {
    planNode = new TopNNode(..., analysis.getLimit(limit.get()), ...);
}
```